native/java/org/jpype/pkg/JPypePackage.java

```diff
@@ -0,0 +1,126 @@
+package org.jpype.pkg;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Map;
+import org.jpype.JPypeKeywords;
+
+public class JPypePackage
+{
+
+  final String pkg;
+  final Map<String, Path> contents;
+
+  public JPypePackage(String pkg, Map<String, Path> contents)
+  {
+    this.pkg = pkg;
+    this.contents = contents;
+  }
+
+  public Object getObject(String name)
+  {
+    Path p = contents.get(name);
+    if (p == null)
+      return null;
+    if (Files.isDirectory(p))
+      return p.toString().replace("/", ".");
+    if (p.toString().endsWith(".class"))
+    {
+      try
+      {
+        if (!isPublic(p))
+          return null;
+        return Class.forName(pkg + "." + JPypeKeywords.unwrap(name));
+      } catch (ClassNotFoundException ex)
+      {
+        return null;
+      }
+    }
+    return null;
+  }
+
+  public String[] getContents()
+  {
+    ArrayList<String> out = new ArrayList<>();
+    for (String key : contents.keySet())
+    {
+      Object o = this.getObject(key);
+      if (o != null)
+        out.add(key);
+    }
+    return out.toArray(new String[out.size()]);
+  }
+
+  static boolean isPublic(Path p)
+  {
+    try (InputStream is = Files.newInputStream(p))
+    {
+      ByteBuffer buffer3 = ByteBuffer.allocate(3);
+
+      // Check the magic
+      ByteBuffer header = ByteBuffer.allocate(4 + 2 + 2 + 2);
+      is.read(header.array());
+      header.rewind();
+      int magic = header.getInt();
+      if (magic != (int) 0xcafebabe)
+        return false;
+      header.getShort(); // skip major
+      header.getShort(); // skip minor
+      int cpitems = header.getShort();
+
+      // Traverse the cp pool
+      for (int i = 0; i < cpitems - 1; ++i)
+      {
+        is.read(buffer3.array());
+        buffer3.rewind();
+        int type = buffer3.get();
```

Comment:
aint't it a byte?

Reply:
Yep.  I am too used to files where we get an int with a value of -1, but this throws instead.